app/src/main/java/com/example/weather/CityPickerAdapter.kt

```diff
@@ -1,52 +1,60 @@
-
+package com.example.weather
 import android.content.Context
 import android.view.LayoutInflater
-import android.view.View
 import android.view.ViewGroup
-import android.widget.TextView
+import androidx.lifecycle.LiveData
 import androidx.recyclerview.widget.RecyclerView
-import com.example.weather.CityPickerFragment
-import com.example.weather.R
+import com.example.weather.databinding.CityItemBinding
+import com.example.weather.network.City
 
 /**
  * Adapter for the [RecyclerView] in [CityPickerFragment].
  */
 
-class CityPickerAdapter(private val dataSet: List<Int>, private val context: Context?) :
+class CityPickerAdapter(private val dataSet: LiveData<List<City>>, private val context: Context?) :
     RecyclerView.Adapter<CityPickerAdapter.ViewHolder>() {
 
     /**
      * Provide a reference to the type of views that you are using
      * (custom ViewHolder)
      */
-    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
-        val textView: TextView
-
-        init {
-            // Define click listener for the ViewHolder's View
-            textView = view.findViewById(R.id.cityTextView)
-        }
-    }
+    class ViewHolder(val binding: CityItemBinding) : RecyclerView.ViewHolder(binding.root)
 
     // Create new views (invoked by the layout manager)
-    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
-        // Create a new view, which defines the UI of the list item
-        val view = LayoutInflater.from(viewGroup.context)
-            .inflate(R.layout.city_item, viewGroup, false)
-
-        return ViewHolder(view)
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+        val binding = CityItemBinding.inflate(inflater, parent, false)
+        return ViewHolder(binding)
     }
 
     // Replace the contents of a view (invoked by the layout manager)
     override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
 
         // Get element from your dataset at this position and replace the
         // contents of the view with that element
-        viewHolder.textView.text = context?.resources?.getString(R.string.list_item, position+1)
+        val dataSetValues: List<City>? = dataSet.value
```

Comment:
if we replace this with something like `val data: City = dataSet.value?.get(position)`, then we can remove all the redundant `dataSetValues?.get(position)?.whatever` calls below and instead do `data?.whatever`. This will make the code look much cleaner